querqy-core/src/main/java/querqy/rewrite/contrib/ReplaceRewriter.java

```diff
@@ -0,0 +1,111 @@
+package querqy.rewrite.contrib;
+
+import querqy.ComparableCharSequence;
+import querqy.CompoundCharSequence;
+import querqy.LowerCaseCharSequence;
+import querqy.model.AbstractNodeVisitor;
+import querqy.model.Clause;
+import querqy.model.DisjunctionMaxQuery;
+import querqy.model.ExpandedQuery;
+import querqy.model.Node;
+import querqy.model.Query;
+import querqy.model.Term;
+import querqy.rewrite.QueryRewriter;
+import querqy.trie.State;
+import querqy.trie.TrieMap;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static querqy.rewrite.contrib.ReplaceRewriterParser.TOKEN_SEPARATOR;
+
+public class ReplaceRewriter extends AbstractNodeVisitor<Node> implements QueryRewriter {
+
+    private final TrieMap<List<ComparableCharSequence>> replaceRules;
+    private final boolean ignoreCase;
+
+    public ReplaceRewriter(TrieMap<List<ComparableCharSequence>> replaceRules, boolean ignoreCase) {
+        this.replaceRules = replaceRules;
+        this.ignoreCase = ignoreCase;
+    }
+
+    private LinkedList<ComparableCharSequence> querySeq = new LinkedList<>();
+    private LinkedList<ComparableCharSequence> matchSeq = new LinkedList<>();
+    private State<List<ComparableCharSequence>> priorMatch = new State<>(false, null, null);
+    private boolean hasReplacement = false;
+
+    @Override
+    public ExpandedQuery rewrite(ExpandedQuery query) {
+        querySeq.clear();
+        matchSeq.clear();
+
+        visit((Query) query.getUserQuery());
+
+        if (priorMatch.isFinal()) {
+            hasReplacement = true;
+            querySeq.addAll(priorMatch.value);
+        } else {
+            if (!matchSeq.isEmpty()) {
+                querySeq.addAll(matchSeq);
+            }
+        }
+
+        return hasReplacement ? buildQueryFromSeqList(querySeq) : query;
+    }
+
+    private ExpandedQuery buildQueryFromSeqList(LinkedList<ComparableCharSequence> tokens) {
+        Query query = new Query();
+        tokens.forEach(token -> {
+            DisjunctionMaxQuery dmq = new DisjunctionMaxQuery(query, Clause.Occur.SHOULD, false);
+            query.addClause(dmq);
+            Term term = new Term(dmq, token);
+            dmq.addClause(term);
+        });
+
```

Comment:
You are creating a new ExpandedQuery from the new 'userQuery'. ExpandedQuery has a few further properties, such as `filterQueries, boostUpQueries, boostDownQueries`, that might have been set by previous rewriters in the chain. Their values will be lost -> please make sure you copy them to the new ExpandedQuery.